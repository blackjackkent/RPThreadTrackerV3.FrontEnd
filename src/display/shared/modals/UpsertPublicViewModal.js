// #region imports
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { ModalHeader, ModalBody, ModalFooter, Button } from 'reactstrap';
import { AvForm } from 'availity-reactstrap-validation';
import { toast } from 'react-toastify';
import TooltipForm from '~/display/forms/TooltipForm';
import Modal from '~/display/shared/styled/Modal';
import { sortCharacters } from '~/utility';
import { useFormReducer } from '~/infrastructure/hooks';
import UpsertPublicViewForm from '~/display/forms/upsert-public-view/UpsertPublicViewForm';
import useUpdatePublicViewMutation from '~/infrastructure/hooks/mutations/public-views/useUpdatePublicViewMutation';
import useCreatePublicViewMutation from '~/infrastructure/hooks/mutations/public-views/useCreatePublicViewMutation';
import LoadingIndicator from '../loading/LoadingIndicator';
// #endregion imports

const propTypes = {
	isModalOpen: PropTypes.bool.isRequired,
	setIsModalOpen: PropTypes.func.isRequired,
	characters: PropTypes.arrayOf(PropTypes.shape({})),
	tags: PropTypes.arrayOf(PropTypes.string),
	columns: PropTypes.shape({}).isRequired,
	actedView: PropTypes.shape({})
};

const UpsertPublicViewModal = (props) => {
	const [publicView, onInputChange, setFormData] = useFormReducer();
	const { createPublicView, isLoading: isCreatePublicViewLoading } =
		useCreatePublicViewMutation();
	const { updatePublicView, isLoading: isUpdatePublicViewLoading } =
		useUpdatePublicViewMutation();
	const isLoading = isCreatePublicViewLoading || isUpdatePublicViewLoading;
	const { actedView, characters, tags, columns, isModalOpen, setIsModalOpen } = props;
	useEffect(() => {
		if (!actedView) {
			return;
		}
		setFormData(actedView);
	}, [setFormData, actedView]);
	const activeCharacters = [].concat(
		characters.sort(sortCharacters).filter((c) => !c.isOnHiatus)
	);

	const submitForm = () => {
		const upsertFn = publicView.id ? updatePublicView : createPublicView;
		upsertFn(publicView)
			.then(() => {
				setIsModalOpen(false);
				toast.success(publicView.id ? 'View updated!' : 'View created!');
			})
			.catch(() => {
				toast.error(
					`There was an error ${publicView.id ? 'updating' : 'creating'} this view.`
				);
			});
	};

	return (
		<Modal isOpen={isModalOpen} toggle={() => setIsModalOpen(!isModalOpen)} backdrop>
			<AvForm onValidSubmit={() => submitForm(publicView)}>
				<ModalHeader toggle={() => setIsModalOpen(!isModalOpen)}>
					{publicView && publicView.id ? 'Edit Public View' : 'Add New Public View'}
				</ModalHeader>
				<ModalBody>
					<TooltipForm
						Renderable={UpsertPublicViewForm}
						publicView={publicView}
						onInputChange={onInputChange}
						characters={activeCharacters}
						tags={tags}
						columns={columns}
					/>
				</ModalBody>
				<ModalFooter>
					{isLoading && <LoadingIndicator />}
					<Button color="primary">
						{publicView.id ? 'Edit Public View' : 'Add Public View'}
					</Button>{' '}
					<Button color="secondary" onClick={() => setIsModalOpen(!isModalOpen)}>
						Cancel
					</Button>
				</ModalFooter>
			</AvForm>
		</Modal>
	);
};

UpsertPublicViewModal.propTypes = propTypes;
UpsertPublicViewModal.defaultProps = {
	characters: [],
	tags: [],
	actedView: {}
};
export default UpsertPublicViewModal;
