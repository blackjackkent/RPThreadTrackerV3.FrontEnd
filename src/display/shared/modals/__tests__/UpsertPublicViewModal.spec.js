// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '~/testhelpers/helpers.unit';
import UpsertPublicViewModal from '../UpsertPublicViewModal';
// #endregion imports

// #region mocks
jest.mock('../../../forms/upsert-public-view/UpsertPublicViewForm', () => 'UpsertPublicViewForm');
jest.mock('../../../forms/TooltipForm', () => 'TooltipForm');
jest.mock('../../../../utility', () => ({
	getValuesFromMultiSelect: (target) => target.multiValues
}));
// #endregion mocks

const createTestProps = (propOverrides) => ({
	isUpsertPublicViewModalOpen: true,
	submitUpsertPublicView: jest.fn(),
	closeUpsertPublicViewModal: jest.fn(),
	viewToEdit: {
		id: '1'
	},
	characters: [
		{
			characterId: 2
		},
		{
			characterId: 3
		}
	],
	tags: ['tag1', 'tag2'],
	columns: {
		TEST_COLUMN: {
			key: 'column1'
		}
	},
	...propOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot if view is new', () => {
			const props = createTestProps({
				viewToEdit: {}
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
});

describe('behavior', () => {
	describe('submitUpsertPublicView', () => {
		it('should be called when form is submitted', () => {
			const submitUpsertPublicView = jest.fn();
			const viewToEdit = {
				id: '2'
			};
			const props = createTestProps({
				submitUpsertPublicView,
				viewToEdit
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith(viewToEdit);
		});
	});
	describe('handleInputChange', () => {
		it('should handle text field update', () => {
			const submitUpsertPublicView = jest.fn();
			const event = {
				target: {
					type: 'text',
					name: 'slug',
					value: 'my-slug'
				}
			};
			const props = createTestProps({
				submitUpsertPublicView
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			element.update();
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith({
				id: '1',
				slug: 'my-slug'
			});
		});
		it('should handle multiselect field update', () => {
			const submitUpsertPublicView = jest.fn();
			const event = {
				target: {
					type: 'select-multiple',
					name: 'columns',
					multiValues: ['value1', 'value2']
				}
			};
			const props = createTestProps({
				submitUpsertPublicView
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			element.update();
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith({
				id: '1',
				columns: ['value1', 'value2']
			});
		});
		it('should handle turnfilter checkbox update', () => {
			const submitUpsertPublicView = jest.fn();
			const event = {
				target: {
					type: 'checkbox',
					name: 'includeMyTurn',
					checked: true
				}
			};
			const props = createTestProps({
				submitUpsertPublicView
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			element.update();
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith({
				id: '1',
				turnFilter: {
					includeMyTurn: true
				}
			});
		});
		it('should handle turnfilter checkbox update if turnfilter is already partially set', () => {
			const submitUpsertPublicView = jest.fn();
			const viewToEdit = {
				id: '1',
				turnFilter: {
					includeTheirTurn: false
				}
			};
			const event = {
				target: {
					type: 'checkbox',
					name: 'includeMyTurn',
					checked: true
				}
			};
			const props = createTestProps({
				submitUpsertPublicView,
				viewToEdit
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			element.instance().handleInputChange(event);
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith({
				id: '1',
				turnFilter: {
					includeMyTurn: true,
					includeTheirTurn: false
				}
			});
		});
	});
	describe('componentWillReceiveProps', () => {
		it('should set view to edit', () => {
			const submitUpsertPublicView = jest.fn();
			const viewToEdit = {
				id: '2'
			};
			const props = createTestProps({
				submitUpsertPublicView
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			element.setProps({
				viewToEdit
			});
			const form = getSpecWrapper(element, 'upsert-public-view-modal-form');
			form.prop('onValidSubmit')();
			expect(submitUpsertPublicView).toHaveBeenCalledTimes(1);
			expect(submitUpsertPublicView).toHaveBeenLastCalledWith(viewToEdit);
		});
	});
	describe('closeUpsertPublicViewModal', () => {
		it('should be triggered when modal is toggled', () => {
			const closeUpsertPublicViewModal = jest.fn();
			const props = createTestProps({
				closeUpsertPublicViewModal
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			const modal = getSpecWrapper(element, 'upsert-public-view-modal');
			modal.prop('toggle')();
			expect(closeUpsertPublicViewModal).toHaveBeenCalledTimes(1);
		});
		it('should be triggered when modal header is toggled', () => {
			const closeUpsertPublicViewModal = jest.fn();
			const props = createTestProps({
				closeUpsertPublicViewModal
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			const header = getSpecWrapper(element, 'upsert-public-view-modal-header');
			header.prop('toggle')();
			expect(closeUpsertPublicViewModal).toHaveBeenCalledTimes(1);
		});
		it('should be triggered when close button is clicked', () => {
			const closeUpsertPublicViewModal = jest.fn();
			const props = createTestProps({
				closeUpsertPublicViewModal
			});
			const jsx = <UpsertPublicViewModal {...props} />;
			const element = shallow(jsx);
			const button = getSpecWrapper(element, 'upsert-public-view-modal-close-button');
			button.simulate('click');
			expect(closeUpsertPublicViewModal).toHaveBeenCalledTimes(1);
		});
	});
});
