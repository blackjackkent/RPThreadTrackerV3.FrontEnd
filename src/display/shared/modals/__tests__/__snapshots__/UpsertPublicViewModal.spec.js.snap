// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<Styled(Modal)
  backdrop={true}
  data-spec="upsert-public-view-modal"
  isOpen={true}
  toggle={[MockFunction]}
>
  <AvForm
    data-spec="upsert-public-view-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      data-spec="upsert-public-view-modal-header"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Edit Public View
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable="UpsertPublicViewForm"
        characters={
          Array [
            Object {
              "characterId": 2,
            },
            Object {
              "characterId": 3,
            },
          ]
        }
        columns={
          Object {
            "TEST_COLUMN": Object {
              "key": "column1",
            },
          }
        }
        handleInputChange={[Function]}
        tags={
          Array [
            "tag1",
            "tag2",
          ]
        }
        viewToEdit={
          Object {
            "id": "1",
          }
        }
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Public View
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-public-view-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Styled(Modal)>
`;

exports[`rendering snapshots should render valid snapshot if view is new 1`] = `
<Styled(Modal)
  backdrop={true}
  data-spec="upsert-public-view-modal"
  isOpen={true}
  toggle={[MockFunction]}
>
  <AvForm
    data-spec="upsert-public-view-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      data-spec="upsert-public-view-modal-header"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Add Public View
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable="UpsertPublicViewForm"
        characters={
          Array [
            Object {
              "characterId": 2,
            },
            Object {
              "characterId": 3,
            },
          ]
        }
        columns={
          Object {
            "TEST_COLUMN": Object {
              "key": "column1",
            },
          }
        }
        handleInputChange={[Function]}
        tags={
          Array [
            "tag1",
            "tag2",
          ]
        }
        viewToEdit={Object {}}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Public View
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-public-view-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Styled(Modal)>
`;
